stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2158 @@
+    /// Finish an operation and update the coordinator selector accordingly
+    fn finish_operation(&mut self) {
+        self.state = State::Idle;
+        self.coordinator_selector.last_message_time = None;
+    }
+
+    /// Update operation
+    fn update_operation(&mut self) {
+        self.state = State::OperationInProgress;
+        self.coordinator_selector.last_message_time = Some(Instant::now());
+    }
+
+    /// Execute the given command and update state accordingly
+    fn execute_command(&mut self, stacks_client: &StacksClient, command: &Command) {
+        match command {
+            Command::Dkg => {
+                //TODO: check if we already have an aggregate key stored in the contract.
+                // If we do, we should not start a new DKG
+                let vote_round = match retry_with_exponential_backoff(|| {
+                    stacks_client
+                        .get_last_round(self.reward_cycle)
+                        .map_err(backoff::Error::transient)
+                }) {
+                    Ok(last_round) => last_round,
+                    Err(e) => {
+                        error!("Signer #{}: Unable to perform DKG. Failed to get last round from stacks node: {e:?}", self.signer_id);
+                        return;
+                    }
+                };
+                // The dkg id will increment internally following "start_dkg_round" so do not increment it here
+                self.coordinator.current_dkg_id = vote_round.unwrap_or(0);
+                info!(
+                    "Signer #{}: Starting DKG vote round {}, for reward cycle {}",
+                    self.signer_id,
+                    self.coordinator.current_dkg_id.wrapping_add(1),
+                    self.reward_cycle
+                );
+                match self.coordinator.start_dkg_round() {
+                    Ok(msg) => {
+                        let ack = self.stackerdb.send_message_with_retry(msg.into());
+                        debug!("Signer #{}: ACK: {ack:?}", self.signer_id);
+                    }
+                    Err(e) => {
+                        error!("Signer #{}: Failed to start DKG: {e:?}", self.signer_id);
+                        return;
+                    }
+                }
+            }
+            Command::Sign {
+                block,
+                is_taproot,
+                merkle_root,
+            } => {
+                let signer_signature_hash = block.header.signer_signature_hash();
+                let block_info = self
+                    .blocks
+                    .entry(signer_signature_hash)
+                    .or_insert_with(|| BlockInfo::new(block.clone()));
+                if block_info.signed_over {
+                    debug!("Signer #{}: Received a sign command for a block we are already signing over. Ignore it.", self.signer_id);
+                    return;
+                }
+                info!("Signer #{}: Signing block: {block:?}", self.signer_id);
```

Comment:
It might be more useful to just log the block's identifying information:

```suggestion
                info!("Signer #{}: Signing block", self.signer_id;
                         "block_consensus_hash" => %block.header.consensus_hash,
                         "block_height" => block.header.chain_length,
                         "pre_sign_block_id" => %block.block_id(),
                );
```